Add Travel Preferences Feature with Full Frontend/Backend Integration

avilotti opened the pull request:
📋 Summary
This PR introduces the Travel Preferences feature where users can define and save personal travel styles, budget range, accommodation preferences, and interests. It integrates with the backend via POST /api/preferences and stores data in MongoDB using the user's ID.

✅ What's Changed
 New page: preferences.html
 POST request to backend with form data
 Integration with preferences.js and MongoDB
 Matching styling via preferences.css
 Link added to navbar for navigation

📎 Related Issues
None directly, but contributes to user personalization features outlined in the proposal.

🔍 How to Test & Screenshots
Navigate to the Preferences page via the navbar
Fill out the form and click "Save Preferences"
Success message will appear
Confirm saved data in MongoDB Atlas under travelpreferences

<img width="1211" height="71" alt="Screenshot 2025-07-29 at 9 01 53 PM" src="https://github.com/user-attachments/assets/c506d77b-4f66-4b06-84bb-4309a0318a51" />
<img width="480" height="569" alt="Screenshot 2025-07-29 at 9 02 01 PM" src="https://github.com/user-attachments/assets/fb0cb0c8-69c3-4a06-ac4b-e19d6ffc8b6f" />
<img width="473" height="617" alt="Screenshot 2025-07-29 at 9 02 18 PM" src="https://github.com/user-attachments/assets/02034939-27cd-48ca-a42e-4f913e1d5ed5" />
<img width="412" height="188" alt="Screenshot 2025-07-29 at 9 02 34 PM" src="https://github.com/user-attachments/assets/c00959a8-c287-4dc3-a24a-bed976250dde" />

🙏 Reviewer Checklist
 Code is readable and commented
 No console errors or warnings
 UI works on desktop and mobile
 All functionality behaves as expected

Ewynman committed:
Develop (#7)

* Add login and fetch user info from the backend (#2)

* Update README.md

* Add PR template

* Add github PR template

* connect frontend to backend and make a test trip API call

* connect to render backend

* initial commit

* Update app config

* Users/ewynman/homepage and auth (#3)

* Add concept for landing page and auth page

* Add Mongo stuff

* Update to ES6 and remove mock (#5)

* Add custom loggers (#6)

Commit:
Resolve merge conflict in createNewTrip

Commit:
Add missing preferences files for frontend feature

Commit:
Add Preferences link to dashboard navbar

Commit:
Fix: Re-add redirect to createtrip.html in createNewTrip

wdsouza0 commented:
I noticed the page wasn't updating the preference fields that the create-trip page was drawing from, so I rewrote some of preferences.js and preferences.html so it should update everything correctly, and refresh the localstorage so any preference changes apply immediately. The preferences page should have parity with create-trip's preferences now:

preferences.js:
`import appConfig from './appConfig.js';

document.addEventListener('DOMContentLoaded', function () {

  const currentUser = localStorage.getItem('currentUser');
    if (!currentUser) {
        // Redirect to login if not authenticated
        window.location.href = './auth.html';
        return;
    }
  const user = JSON.parse(currentUser);

  if (user.travelPreferences.travelStyle) {
    const travelStyleSelect = document.getElementById('select-travel-style');
    travelStyleSelect.value = user.travelPreferences.travelStyle;
  }

  if (user.travelPreferences.accommodationStyle) {
    const travelAccommodationSelect = document.getElementById('select-accommodation-style');
    travelAccommodationSelect.value = user.travelPreferences.accommodationStyle;
  }

  if (user.travelPreferences.budgetRange) {
    const travelBudgetSelect = document.getElementById('select-budget-range');
    travelBudgetSelect.value = user.travelPreferences.budgetRange;
  }

  user.travelPreferences.interests.forEach(interest => {
      //console.log(interest);
      const checkbox = document.querySelector(`#select-interests input[type="checkbox"][value="${interest}"]`);
      if (checkbox) {
        checkbox.checked = true;
      }
    });

})


document.getElementById('preferencesForm').addEventListener('submit', async (e) => {
  e.preventDefault();

  const form = e.target;
  
  const currentUser = localStorage.getItem('currentUser');
    if (!currentUser) {
        // Redirect to login if not authenticated
        window.location.href = './auth.html';
        return;
    }
  const user = JSON.parse(currentUser);
  const userId = user._id
  //console.log(userId)
  const data = {
    travelPreferences: {
      travelStyle: form.travelStyle.value,
      budgetRange: form.budgetRange.value,
      accommodationStyle: form.accommodationStyle.value,
      interests: Array.from(form.querySelectorAll('#select-interests input[type="checkbox"]:checked')).map(i => i.value)
    }
  };

/*  .then(...) implementation:
    fetch(`${appConfig.API_BASE}/users/profile/${userId}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    })          
    .then(res => {
      if (!res.ok) {
        return res.json().then(err => {
          throw new Error(err.error || 'Failed to update user preferences');
        });
      }
      return res.json();
    })
    .then(data => {
      document.getElementById('successMsg').style.display = 'block';
      console.log('✅ Preferences saved:', data); 
      //Refresh localstorage so new preferences apply
      localStorage.setItem('currentUser', JSON.stringify(data));
  })
    .catch(err => {
      console.log('Update failed: ' + err.message, 'error');
      alert(err.message || 'Failed to save preferences.');
    });
*/
    const response = await fetch(`${appConfig.API_BASE}/users/profile/${userId}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data)
    });
    
    const result = await response.json();

    if(!response.ok)
    {
          throw new Error(err.error || 'Failed to update user preferences');
    }
    else
    {
      document.getElementById('successMsg').style.display = 'block';
      //Refresh localstorage so new preferences apply to the create a trip page
      localStorage.setItem('currentUser', JSON.stringify(result));
    }
})`

preferences.html:
`<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Travel Preferences | Travel Buddy</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />

    <!-- Google Fonts -->
    <link rel="preconnect" href="https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
    <link
      href="https://fonts.googleapis.com/css2?family=Inter:wght@400;600;700&family=Playfair+Display:wght@700&display=swap"
      rel="stylesheet"
    />

    <!-- Font Awesome -->
    <link
      rel="stylesheet"
      href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.0/css/all.min.css"
    />

    <!-- CSS -->
    <link rel="stylesheet" href="../CSS/index.css" />
    <link rel="stylesheet" href="../CSS/preferences.css" />
  </head>
  <body>
    <!-- MATCHING NAVBAR -->
    <nav class="navbar">
      <div class="nav-container">
        <div class="nav-brand">
          <i class="fas fa-compass nav-icon"></i>
          <span>Travel Buddy</span>
        </div>
        <div class="nav-menu">
          <a href="../index.html" class="nav-link">Home</a>
          <a href="#" class="nav-link">Features</a>
          <a href="#" class="nav-link">Reviews</a>
          <a href="dashboard.html" class="nav-link">Dashboard</a>
          <a href="profile.html" class="nav-link">Profile</a>
          <a href="preferences.html" class="nav-link">Preferences</a>
          <a href="#" class="nav-link nav-cta">Logout</a>
        </div>
      </div>
    </nav>

    <!-- TRAVEL PREFERENCES FORM -->
    <main class="preferences-page">
      <div class="preferences-card">
        <h2>Travel Preferences</h2>
        <form id="preferencesForm">
          <label>Travel Style</label>
          <select name="travelStyle" id="select-travel-style" required>
            <option value="relaxed">Relaxed</option>
            <option value="balanced">Balanced</option>
            <option value="fast-paced">Fast-paced</option>
          </select>

          <label>Budget Range</label>
          <select name="budgetRange" id="select-budget-range" required>
            <option value="budget">Budget</option>
            <option value="mid-range">Mid-range</option>
            <option value="luxury">Luxury</option>
          </select>

          <label>Accommodation Style</label>
          <select name="accommodationStyle" id="select-accommodation-style" required>
            <option value="hotels">Hotels</option>
            <option value="hostels">Hostels</option>
            <option value="vacation-rentals">Vacation Rentals</option>
            <option value="eco-lodges">Eco-lodges</option>
          </select>

          <label>Interests</label>
          <div class="checkbox-group" id="select-interests">
              <label><input type="checkbox" value="food"> Food</label>
              <label><input type="checkbox" value="culture"> Culture</label>
              <label><input type="checkbox" value="adventure"> Adventure</label>
              <label><input type="checkbox" value="nature"> Nature</label>
              <label><input type="checkbox" value="nightlife"> Nightlife</label>
              <label><input type="checkbox" value="history"> History</label>
              <label><input type="checkbox" value="wellness"> Wellness</label>
              <label><input type="checkbox" value="photography"> Photography</label>
              <label><input type="checkbox" value="shopping"> Shopping</label>
              <label><input type="checkbox" value="sports"> Sports</label>
              <label><input type="checkbox" value="art"> Art</label>
              <label><input type="checkbox" value="music"> Music</label>
              <label><input type="checkbox" value="architecture"> Architecture</label>
              <label><input type="checkbox" value="beaches"> Beaches</label>
              <label><input type="checkbox" value="mountains"> Mountains</label>
              <label><input type="checkbox" value="cities"> Cities</label>
              <label><input type="checkbox" value="rural"> Rural</label>
              <label><input type="checkbox" value="temples"> Temples</label>
              <label><input type="checkbox" value="museums"> Museums</label>
              <label><input type="checkbox" value="local-experiences"> Local Experiences</label>
          </div>

          <button type="submit" class="auth-btn">Save Preferences</button>
        </form>
        <p id="successMsg" style="display: none;">Preferences saved!</p>
      </div>
    </main>

    <script type="module" src="../JS/preferences.js"></script>
  </body>
</html>
`

wdsouza0 committed:
Updated preferences.js for parity with createtrip.html/js

preferences.js now updates the DB entries that createtrip.js uses, and updates the localstorage. preferences.html now includes all preferences options that createtrip.html has

wdsouza0 commented:
I noticed the page wasn't updating the preference fields that the create-trip page was drawing from, so I rewrote some of preferences.js and preferences.html so it should update everything correctly, and refresh the localstorage so any preference changes apply immediately. The preferences page should also have parity with create-trip's preferences now. I pushed the changes in a new commit, so lmk if theres any problems and I can revert it

wdsouza0 committed:
Use logger instead of console
